awesome/src/common/object.rs

```diff
@@ -1,32 +1,59 @@
 //! Utility methods and constructors for Lua objects
 
+use std::convert::From;
+use std::fmt::{self, Display, Formatter};
+use std::cell;
+use std::marker::PhantomData;
+
+use rlua::{self, AnyUserData, FromLua, Function, Lua, MetaMethod, Table, ToLua, UserData,
+           UserDataMethods, Value};
+
 use super::class::Class;
 use super::property::Property;
 use super::signal;
-use rlua::{self, AnyUserData, FromLua, Function, Lua, MetaMethod, Table, ToLua, UserData,
-           UserDataMethods, Value};
-use std::convert::From;
-use std::fmt::Display;
+
+/// Define the traits for states
+pub trait State: UserData + Default + Display + Send {}
+impl<T> State for T where T: UserData + Default + Display + Send {}
 
 /// All Lua objects can be cast to this.
-#[derive(Clone, Debug)]
-pub struct Object<'lua> {
-    pub object: AnyUserData<'lua>
+#[derive(Debug)]
+pub struct Object<'lua, S: State>{
+    pub obj: AnyUserData<'lua>,
+    state: PhantomData<S>,
+}
+
+impl<'lua, S: State> Clone for Object<'lua, S> {
+    fn clone(&self) -> Self {
+        Object { obj: self.obj.clone(), state: PhantomData }
+    }
+}
+
+impl<'lua, S: State> From<AnyUserData<'lua>> for Object<'lua, S> {
+    fn from(obj: AnyUserData<'lua>) -> Self {
+        Object { obj, state: PhantomData }
+    }
+}
+
+impl<'lua, S: State> Into<AnyUserData<'lua>> for Object<'lua, S> {
+    fn into(self) -> AnyUserData<'lua> {
+        self.obj
+    }
 }
 
-impl<'lua> From<AnyUserData<'lua>> for Object<'lua> {
-    fn from(object: AnyUserData<'lua>) -> Self {
-        Object { object }
+impl<'lua, S: State> Display for Object<'lua, S> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Object: {:p}", self)
```

Comment:
The previous way means that one does not have to implement a redundant Display method for the various State types though. The only way it is useful(-ish) to display a pointer is with the lua facing side, but the name property is already there. I think this would incur a higher code bloat without much benefit, so I propose removing both this and the object `Display` methods.

Reply:
The object implements `__tostring` by using the results of the `Display` trait. This is based on the [C implementation](https://github.com/awesomeWM/awesome/blob/master/common/luaobject.c#L334).

Reply:
But it could easily use `default_tostring` with the same results if I am not mistaken.

Reply:
Ah so it could. Great, then lets delete the `Display` trait requirement and just use that.